### PR TITLE
Support config domain URL for cloudcore

### LINKED ISSUE #1642: Add advertise address for cloudcore
We have added the feature of certificate issuance in kubeedge, and now we need to add a filed `advertise-address` to specify the service address of the cloudcore, which will be used in the SANs of the cloudcore certificate.

And now in cloudhub, websocket/quic all have [Address](https://github.com/kubeedge/kubeedge/blob/master/pkg/apis/componentconfig/cloudcore/v1alpha1/types.go#L117) field, when we add the `advertise -address`, is the `Address` in websocket/quic still needed?

The `advertise -address` will be a specific address like 1.2.3.4, but  `Address` in websocket/quic now default as `0.0.0.0`.

/cc @kevin-wangzefeng @kadisi @chendave @subpathdev  @GsssC @ls889 @XJangel 

Task List:
- [x] Add advertise address in cloudcore/cloudhub #1643
- [x] Update the docs for the usage of advertise address #1656 
- [ ] Add bind-address config option and remove address in websocket/quic/httpserver
- [x] Support config domain URL for cloudcore #2126

### PR DESCRIPTION
**Pull request by ls889**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #1642 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by kevin-wangzefeng** [comment]
Removed irrelevant labels.

File: cloud/pkg/cloudhub/servers/httpserver/signcerts.go [hunk 1]
```diff
@@ -39,7 +39,8 @@ func SignCerts() ([]byte, []byte, error) {
 		Organization: []string{"KubeEdge"},
 		Usages:       []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
 		AltNames: certutil.AltNames{
-			IPs: getIps(hubconfig.Config.AdvertiseAddress),
+			DNSNames: hubconfig.Config.DNSNames,
```
Comment (kevin-wangzefeng) [inline cloud/pkg/cloudhub/servers/httpserver/signcerts.go:42]: nit: DNSNames => DomainNames ?

Comment (ls889) [reply cloud/pkg/cloudhub/servers/httpserver/signcerts.go:42]: > nit: DNSNames => DomainNames ?

@kevin-wangzefeng Thank for your suggestion. DNSNames is a bit weird, but in k8s, it is named like this, so we'd better keep using DNSNames.

File: keadm/cmd/keadm/app/cmd/util/cloudinstaller.go [hunk 2]
```diff
@@ -63,6 +64,10 @@ func (cu *KubeCloudInstTool) InstallTools() error {
 			cloudCoreConfig.Modules.CloudHub.AdvertiseAddress = strings.Split(cu.AdvertiseAddress, ",")
 		}
 
+		if cu.DNSName != "" {
+			cloudCoreConfig.Modules.CloudHub.DNSNames = strings.Split(cu.DNSName, ",")
```
Comment (kevin-wangzefeng) [inline keadm/cmd/keadm/app/cmd/util/cloudinstaller.go:68]: We need validations somewhere on the user input

Comment (ls889) [reply keadm/cmd/keadm/app/cmd/util/cloudinstaller.go:68]: @kevin-wangzefeng  sure, I will be create a new pr to do it.

File: pkg/apis/componentconfig/cloudcore/v1alpha1/types.go [hunk 3]
```diff
@@ -114,6 +115,8 @@ type CloudHub struct {
 	HTTPS *CloudHubHTTPS `json:"https,omitempty"`
 	// AdvertiseAddress sets the IP address for the cloudcore to advertise.
 	AdvertiseAddress []string `json:"advertiseAddress,omitempty"`
+	// DNSNames sets the DNSNames for CloudCore.
+	DNSNames []string `json:"DNSNames,omitempty"`
```
Comment (fisherxu) [inline pkg/apis/componentconfig/cloudcore/v1alpha1/types.go:119]:
```suggestion
	DNSNames []string `json:"dnsNames,omitempty"`
```

File: pkg/apis/componentconfig/cloudcore/v1alpha1/default.go [hunk 4]
```diff
@@ -184,6 +185,7 @@ func NewMinCloudCoreConfig() *CloudCoreConfig {
 				TLSCertFile:       constants.DefaultCertFile,
 				TLSPrivateKeyFile: constants.DefaultKeyFile,
 				AdvertiseAddress:  []string{advertiseAddress.String()},
+				DNSNames:          []string{""},
```
Comment (fisherxu) [inline pkg/apis/componentconfig/cloudcore/v1alpha1/default.go:188]: No need add it to min and default config.

File: keadm/cmd/keadm/app/cmd/common/types.go [hunk 5]
```diff
@@ -28,6 +28,7 @@ type InitOptions struct {
 	KubeConfig       string
 	Master           string
 	AdvertiseAddress string
+	DNS              string
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/common/types.go:31]: Does the `keadm` command has the `DNS` flag option?

Comment (ls889) [reply keadm/cmd/keadm/app/cmd/common/types.go:31]: @fisherxu  yes, in another pr.

Comment (fisherxu) [reply keadm/cmd/keadm/app/cmd/common/types.go:31]: We can add it to this one.

File: pkg/apis/componentconfig/cloudcore/v1alpha1/default.go (hunk 4 above)
Comment (ls889) [reply pkg/apis/componentconfig/cloudcore/v1alpha1/default.go:188]: @fisherxu So, we want users to write this field in yaml manually?

File: keadm/cmd/keadm/app/cmd/common/types.go (hunk 5 above)
Comment (ls889) [reply keadm/cmd/keadm/app/cmd/common/types.go:31]: @fisherxu  ok, i will merge them later.

File: keadm/cmd/keadm/app/cmd/cloud/init.go [hunk 6]
```diff
@@ -97,6 +97,9 @@ func addJoinOtherFlags(cmd *cobra.Command, initOpts *types.InitOptions) {
 
 	cmd.Flags().StringVar(&initOpts.AdvertiseAddress, types.AdvertiseAddress, initOpts.AdvertiseAddress,
 		"Use this key to set SANs in certificate of cloudcore. eg: 10.10.102.78,10.10.102.79")
+
+	cmd.Flags().StringVar(&initOpts.DNS, types.DomainName, initOpts.DNS,
+		"Use this key to set SANs in certificate of cloudcore. eg: www.cloudcore.cn,www.kubeedge.cn")
```
Comment (fisherxu) [inline keadm/cmd/keadm/app/cmd/cloud/init.go:102]:
```suggestion
		"Use this key to set domain names in cloudcore's certificate SubAltNames field. eg: www.cloudcore.cn,www.kubeedge.cn")
```

Also update the ip field above.

File: pkg/apis/componentconfig/cloudcore/v1alpha1/default.go (hunk 4 above)
Comment (fisherxu) [reply pkg/apis/componentconfig/cloudcore/v1alpha1/default.go:188]: Sounds good

File: keadm/cmd/keadm/app/cmd/cloud/init.go (hunk 6 above)
Comment (ls889) [reply keadm/cmd/keadm/app/cmd/cloud/init.go:102]: @fisherxu  ok

File: pkg/apis/componentconfig/cloudcore/v1alpha1/default.go (hunk 4 above)
Comment (ls889) [reply pkg/apis/componentconfig/cloudcore/v1alpha1/default.go:188]: @fisherxu  I will remove it right now.

**Comment by fisherxu** [comment]
/lgtm
/approve
Please raise another PR to fix the comment https://github.com/kubeedge/kubeedge/pull/2126#discussion_r483350323

**Pull request merged**